### DISCUSSION TITLE
ServerPrerendered

### DISCUSSION THREAD

**Discussion by KieranFleckney** [DavidVollmers/Ignis#49] [General]
I am trying add ignis components to the new .net 8 app template. I understand from the docs https://ignis.dvolper.dev/docs/components/lifecycle that components using the ignis base don't re-render twice when using pre-rendering. However,  I want pre-rendering to work, as from what I can tell ignis component don't pre-render at all. In the docs it says to implement IHandleAfterRender to get this desires result. However, I can not understand how that would work or get it to work. Any guidance would be good. Or would is it recommend that any component that needs pre-rendering to not be a ignis component

**Comment by DavidVollmers** [comment]
Yes documentation doesn't make sense here. Currently you cannot handle prerendering if you inherit the `IgnisComponentBase`. I already have an idea on how to enable that easily. I should be able to deliver it today or at least this week.

**Reply by KieranFleckney** [reply]
No rush, makes me feel bad, sounds like I'm giving you work to do 😅  I don't need it currently, however, it would be nice to have the option for it. Maybe worthing creating a issue to track it?

**Reply by DavidVollmers** [reply]
I already created an isssue. And don't worry I am not rushing, but I already have a good idea on how to solve this. Also if you want people to use your rather unknown framework you need to go the extra mile 😄

**Comment by DavidVollmers** [comment]
With version `1.3.0` the `PrerenderAttribute` got introduced to enable prerendering for a component. The documentation got updated accordingly: https://ignis.dvolper.dev/docs/components/lifecycle

**Comment by marshalhayes** [comment]
If I wanted to prerender all components by default, would adding this attribute on both the server & client `_Imports.razor` work?

```cs
// WebApp _Imports.razor
@attribute [Prerender]
```

```cs
// WebApp.Client _Imports.razor
@attribute [Prerender]
```

This was a hurdle for me (albeit a small one) when I was adding Ignis to my project. I was surprised to find I needed to opt-in to prerendering since [prerendering is enabled by default](https://learn.microsoft.com/en-us/aspnet/core/blazor/components/render-modes?view=aspnetcore-9.0#prerendering) in Blazor.

**Reply by DavidVollmers** [reply]
So, on server side this should work but on client (if you mean web assembly) there is no prerendering cycle, but it should not matter there.

https://github.com/DavidVollmers/Ignis/blob/master/packages/Ignis.Components.Server/ServerHostContext.cs
https://github.com/DavidVollmers/Ignis/blob/master/packages/Ignis.Components.WebAssembly/WebAssemblyHostContext.cs
https://github.com/DavidVollmers/Ignis/blob/master/packages/Ignis.Components/IgnisComponentBase.cs#L43
